File: WQChartsDemo/Objc/View/CheckboxCell.m
// https://github.com/CoderWQYao/WQCharts-iOS
//
// CheckboxCell.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "CheckboxCell.h"

@implementation CheckboxCell

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.style = ButtonsCellStyleCheckBox;
    }
    return self;
}

- (NSArray<NSNumber *> *)selections {
    NSMutableArray<NSNumber *>* selections = [NSMutableArray array];
    for (NSInteger i=0; i<self.buttonsWrap.subviews.count; i++) {
        UIButton* button = self.buttonsWrap.subviews[i];
        if(button.isSelected) {
            [selections addObject:@(i)];
        }
    }
    return selections;
}

- (void)setSelections:(NSArray<NSNumber *> *)selections {
    for (NSInteger i=0; i<self.buttonsWrap.subviews.count; i++) {
        UIButton* button = self.buttonsWrap.subviews[i];
        [self adjustButtonStyle:button isSelected:NO];
    }
    
    for (NSInteger i=0; i<selections.count; i++) {
        NSInteger selection = [selections[i] integerValue];
        [self adjustButtonStyle:self.buttonsWrap.subviews[selection] isSelected:YES];
    }
}

- (CheckboxCell*(^)(NSString*))setTitle {
    return ^CheckboxCell *(NSString *title) {
        self.title = title;
        return self;
    };
}

- (CheckboxCell*(^)(NSArray<NSString*>*))setOptions {
    return ^CheckboxCell *(NSArray<NSString *>* options) {
        self.options = options;
        return self;
    };
}

- (CheckboxCell * _Nonnull (^)(NSArray<NSNumber *> * _Nonnull))setSelections {
    return ^CheckboxCell *(NSArray<NSNumber *> * selections) {
        self.selections = selections;
        return self;
    };
}

- (CheckboxCell*(^)(void (^)(CheckboxCell*, NSArray<NSNumber*>*)))setOnSelectionsChange {
    return ^CheckboxCell *(void (^onSelectionsChange)(CheckboxCell*, NSArray<NSNumber*>*)) {
        self.onSelectionsChange = onSelectionsChange;
        return self;
    };
}


- (void)callSectionsChange {
    if (self.onSelectionsChange) {
        self.onSelectionsChange(self, self.selections);
    }
}

- (void)rebuildButtons {
    NSArray<NSNumber*>* selections = self.selections;
    
    [super rebuildButtons];
    
    UIView* buttonsWrap = self.buttonsWrap;
    for (UIButton* button in buttonsWrap.subviews) {
        [button addTarget:self action:@selector(handleButtonClick:) forControlEvents:UIControlEventTouchUpInside];
    }
    
    BOOL selectionsChanged = NO;
    NSInteger buttonCount = buttonsWrap.subviews.count;
    for (NSInteger i = 0; i<selections.count; i++) {
        NSInteger selection = [selections[i] integerValue];
        if (selection < 0 || selection > buttonCount - 1) {
            selectionsChanged = YES;
        } else {
            [self adjustButtonStyle:buttonsWrap.subviews[selection] isSelected:YES];
        }
    }
    
    if (selectionsChanged) {
        [self callSectionsChange];
    }
}


- (void)handleButtonClick:(UIButton *)sender {
    [self adjustButtonStyle:sender isSelected:!sender.selected];
    [self callSectionsChange];
}

@end

File: WQChartsDemo/Swift/View/CheckboxCell.swift
// https://github.com/CoderWQYao/WQCharts-iOS
//
// CheckboxCell.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class CheckboxCell: ButtonsCell {
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.style = .Checkbox
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    open var selections: [Int] {
        get {
            let selections = NSMutableArray()
            for i in 0..<buttonsWrap.subviews.count {
                let button = buttonsWrap.subviews[i] as! UIButton
                if button.isSelected {
                    selections.add(i)
                }
            }
            return selections as! [Int]
        }
        set {
            for i in 0..<buttonsWrap.subviews.count {
                let button = buttonsWrap.subviews[i] as! UIButton
                adjustButtonStyle(button, false)
            }
            
            for selection in newValue {
                adjustButtonStyle(buttonsWrap.subviews[selection] as! UIButton, true)
            }
        }
    }
    
    
    open func setTitle(_ title: String) -> CheckboxCell {
        self.title = title
        return self
    }
    
    open func setOptions(_ options: [String]?) -> CheckboxCell {
        self.options = options
        return self
    }
    
    open func setSelections(_ selections: [Int]) -> CheckboxCell {
        self.selections = selections
        return self
    }
    
    open var onSelectionsChange: ((CheckboxCell, [Int]) -> Void)?
    
    open func setOnSelectionsChange(_ onSelectionsChange: ((CheckboxCell, [Int]) -> Void)?) -> CheckboxCell {
        self.onSelectionsChange = onSelectionsChange
        return self
    }
    
    open func callSectionsChange() {
        if let onSelectionsChange = onSelectionsChange {
            onSelectionsChange(self, selections)
        }
    }
    
    override func rebuildButtons() {
        let selections = self.selections
        
        super.rebuildButtons()
        
        let buttonsWrap = self.buttonsWrap
        for button in buttonsWrap.subviews as! [UIButton] {
            button.addTarget(self, action: #selector(handleButtonClick), for: .touchUpInside)
        }
        
        var selectionsChanged = false
        let buttonCount: Int = buttonsWrap.subviews.count
        for i in 0..<selections.count {
            let selection = selections[i]
            if selection < 0 || selection > buttonCount - 1 {
                selectionsChanged = true
            } else {
                adjustButtonStyle(buttonsWrap.subviews[selection] as! UIButton, true)
            }
        }
        
        if selectionsChanged {
            callSectionsChange()
        }
    }
    
    @objc func handleButtonClick(_ sender: UIButton) {
        adjustButtonStyle(sender, !sender.isSelected)
        callSectionsChange()
    }
    
}
